servant-server/src/Servant/Server/Internal/ErrorFormatter.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE DataKinds #-}
+
+-- | TODO: more documentation for this module
+module Servant.Server.Internal.ErrorFormatter
+  where
+
+import           Data.String.Conversions
+                 (cs)
+import           Data.Typeable
+
+import           Servant.API.ContentTypes
+import           Servant.Server.Internal.Context
+import           Servant.Server.Internal.ServerError
+
+-- | 'Context' that contains default formatters for all error types.
+--
+-- Default formatters will just return HTTP 400 status code with error
+-- message as response body.
+type DefaultErrorFormatters = '[BodyParseErrorFormatter, URLParseErrorFormatter, HeaderParseErrorFormatter]
+
+defaultErrorFormatters :: Context DefaultErrorFormatters
+defaultErrorFormatters =
+  defaulyBodyParseErrorFormatter
+  :. defaultURLParseErrorFormatter
+  :. defaultHeaderParseErrorFormatter
+  :. EmptyContext
+
+-- | A custom formatter for errors produced by parsing combinators like
+-- 'Servant.API.ReqBody' or 'Servant.API.Capture'.
+--
+-- A 'TypeRep' argument described the concrete combinator that raised
+-- the error, allowing formatter to customize the message for different
+-- combinators.
+type ErrorFormatter = TypeRep -> AcceptHeader -> String -> ServerError
```

Comment:
I think the errors showing which combinator threw them is really cool! It has the potential to make a lot of debugging much easier!

Reply:
I'm not sure about the rest of this though. We now have `AcceptHeader`; next release we might want `AcceptLanguage` (so you can do localized errors); release after that something else. Each time, it's going to be a breaking change.

Would `Request` not maybe be better? If there's a reason not to go with that, then maybe a new datatype that looks something like:

```
data ErrorContext = ErrorContext 
   { requestAcceptHeader :: AcceptHeader
   , throwingCombinator :: TypeRep
   , errorMessage :: String
   } 
```
(And then `ErrorFormatter = ErrorContext -> ServerError`).

Note that if there is such a datatype, the constructor should only be exported from an `Internal` module; from eg `Servant` itself, only the field accessors should be exported (that way, we can stay backward compatible while adding new fields).

Reply:
Hmm, actually I think `Request` is the best way to go, I'll make the change.